File: lib/swift/Sources/TProtocolDecorator.swift
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import Foundation

class TProtocolDecorator: TProtocol {

  private let proto: TProtocol
  var transport: TTransport

  init(proto: TProtocol) {
    self.proto = proto
    self.transport = proto.transport
  }

  required init(on transport: TTransport) {
    fatalError("init(on:) has not been implemented")
  }

  func readMessageBegin() throws -> (String, TMessageType, Int32) {
    return try proto.readMessageBegin()
  }

  func readMessageEnd() throws {
    try proto.readMessageEnd()
  }

  func readStructBegin() throws -> String {
    return try proto.readStructBegin()
  }

  func readStructEnd() throws {
    try proto.readStructEnd()
  }

  func readFieldBegin() throws -> (String, TType, Int32) {
    return try proto.readFieldBegin()
  }

  func readFieldEnd() throws {
    try proto.readFieldEnd()
  }

  func readMapBegin() throws -> (TType, TType, Int32) {
    return try proto.readMapBegin()
  }

  func readMapEnd() throws {
    try proto.readMapEnd()
  }

  func readSetBegin() throws -> (TType, Int32) {
    return try proto.readSetBegin()
  }

  func readSetEnd() throws {
    try proto.readSetEnd()
  }

  func readListBegin() throws -> (TType, Int32) {
    return try proto.readListBegin()
  }

  func readListEnd() throws {
    try proto.readListEnd()
  }

  func read() throws -> String {
    return try proto.read()
  }

  func read() throws -> Bool {
    return try proto.read()
  }

  func read() throws -> UInt8 {
    return try proto.read()
  }

  func read() throws -> Int16 {
    return try proto.read()
  }

  func read() throws -> Int32 {
    return try proto.read()
  }

  func read() throws -> Int64 {
    return try proto.read()
  }

  func read() throws -> Double {
    return try proto.read()
  }

  func read() throws -> Data {
    return try proto.read()
  }

  func writeMessageBegin(name: String, type messageType: TMessageType, sequenceID: Int32) throws {
    try proto.writeMessageBegin(name: name, type: messageType, sequenceID: sequenceID)
  }

  func writeMessageEnd() throws {
    try proto.writeMessageEnd()
  }

  func writeStructBegin(name: String) throws {
    try proto.writeStructBegin(name: name)
  }

  func writeStructEnd() throws {
    try proto.writeStructEnd()
  }

  func writeFieldBegin(name: String, type fieldType: TType, fieldID: Int32) throws {
    try proto.writeFieldBegin(name: name, type: fieldType, fieldID: fieldID)
  }

  func writeFieldStop() throws {
    try proto.writeFieldStop()
  }

  func writeFieldEnd() throws {
    try proto.writeFieldEnd()
  }

  func writeMapBegin(keyType: TType, valueType: TType, size: Int32) throws {
    try proto.writeMapBegin(keyType: keyType, valueType: valueType, size: size)
  }

  func writeMapEnd() throws {
    try proto.writeMapEnd()
  }

  func writeSetBegin(elementType: TType, size: Int32) throws {
    try proto.writeSetBegin(elementType: elementType, size: size)
  }

  func writeSetEnd() throws {
    try proto.writeSetEnd()
  }

  func writeListBegin(elementType: TType, size: Int32) throws {
    try proto.writeListBegin(elementType: elementType, size: size)
  }

  func writeListEnd() throws {
    try proto.writeListEnd()
  }

  func write(_ value: String) throws {
    try proto.write(value)
  }

  func write(_ value: Bool) throws {
    try proto.write(value)
  }

  func write(_ value: UInt8) throws {
    try proto.write(value)
  }

  func write(_ value: Int16) throws {
    try proto.write(value)
  }

  func write(_ value: Int32) throws {
    try proto.write(value)
  }

  func write(_ value: Int64) throws {
    try proto.write(value)
  }

  func write(_ value: Double) throws {
    try proto.write(value)
  }

  func write(_ value: Data) throws {
    try proto.write(value)
  }
}

File: lib/java/src/main/java/org/apache/thrift/protocol/TProtocolDecorator.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.thrift.protocol;

import java.nio.ByteBuffer;
import java.util.UUID;
import org.apache.thrift.TException;

/**
 * <code>TProtocolDecorator</code> forwards all requests to an enclosed <code>TProtocol</code>
 * instance, providing a way to author concise concrete decorator subclasses. While it has no
 * abstract methods, it is marked abstract as a reminder that by itself, it does not modify the
 * behaviour of the enclosed <code>TProtocol</code>.
 *
 * <p>See p.175 of Design Patterns (by Gamma et al.)
 *
 * @see org.apache.thrift.protocol.TMultiplexedProtocol
 */
public abstract class TProtocolDecorator extends TProtocol {

  private final TProtocol concreteProtocol;

  /**
   * Encloses the specified protocol.
   *
   * @param protocol All operations will be forward to this protocol. Must be non-null.
   */
  public TProtocolDecorator(TProtocol protocol) {
    super(protocol.getTransport());
    concreteProtocol = protocol;
  }

  @Override
  public void writeMessageBegin(TMessage tMessage) throws TException {
    concreteProtocol.writeMessageBegin(tMessage);
  }

  @Override
  public void writeMessageEnd() throws TException {
    concreteProtocol.writeMessageEnd();
  }

  @Override
  public void writeStructBegin(TStruct tStruct) throws TException {
    concreteProtocol.writeStructBegin(tStruct);
  }

  @Override
  public UUID readUuid() throws TException {
    return concreteProtocol.readUuid();
  }

  @Override
  public void writeUuid(UUID uuid) throws TException {
    concreteProtocol.writeUuid(uuid);
  }

  @Override
  public void writeStructEnd() throws TException {
    concreteProtocol.writeStructEnd();
  }

  @Override
  public void writeFieldBegin(TField tField) throws TException {
    concreteProtocol.writeFieldBegin(tField);
  }

  @Override
  public void writeFieldEnd() throws TException {
    concreteProtocol.writeFieldEnd();
  }

  @Override
  public void writeFieldStop() throws TException {
    concreteProtocol.writeFieldStop();
  }

  @Override
  public void writeMapBegin(TMap tMap) throws TException {
    concreteProtocol.writeMapBegin(tMap);
  }

  @Override
  public void writeMapEnd() throws TException {
    concreteProtocol.writeMapEnd();
  }

  @Override
  public void writeListBegin(TList tList) throws TException {
    concreteProtocol.writeListBegin(tList);
  }

  @Override
  public void writeListEnd() throws TException {
    concreteProtocol.writeListEnd();
  }

  @Override
  public void writeSetBegin(TSet tSet) throws TException {
    concreteProtocol.writeSetBegin(tSet);
  }

  @Override
  public void writeSetEnd() throws TException {
    concreteProtocol.writeSetEnd();
  }

  @Override
  public void writeBool(boolean b) throws TException {
    concreteProtocol.writeBool(b);
  }

  @Override
  public void writeByte(byte b) throws TException {
    concreteProtocol.writeByte(b);
  }

  @Override
  public void writeI16(short i) throws TException {
    concreteProtocol.writeI16(i);
  }

  @Override
  public void writeI32(int i) throws TException {
    concreteProtocol.writeI32(i);
  }

  @Override
  public void writeI64(long l) throws TException {
    concreteProtocol.writeI64(l);
  }

  @Override
  public void writeDouble(double v) throws TException {
    concreteProtocol.writeDouble(v);
  }

  @Override
  public void writeString(String s) throws TException {
    concreteProtocol.writeString(s);
  }

  @Override
  public void writeBinary(ByteBuffer buf) throws TException {
    concreteProtocol.writeBinary(buf);
  }

  @Override
  public TMessage readMessageBegin() throws TException {
    return concreteProtocol.readMessageBegin();
  }

  @Override
  public void readMessageEnd() throws TException {
    concreteProtocol.readMessageEnd();
  }

  @Override
  public TStruct readStructBegin() throws TException {
    return concreteProtocol.readStructBegin();
  }

  @Override
  public void readStructEnd() throws TException {
    concreteProtocol.readStructEnd();
  }

  @Override
  public TField readFieldBegin() throws TException {
    return concreteProtocol.readFieldBegin();
  }

  @Override
  public void readFieldEnd() throws TException {
    concreteProtocol.readFieldEnd();
  }

  @Override
  public TMap readMapBegin() throws TException {
    return concreteProtocol.readMapBegin();
  }

  @Override
  public void readMapEnd() throws TException {
    concreteProtocol.readMapEnd();
  }

  @Override
  public TList readListBegin() throws TException {
    return concreteProtocol.readListBegin();
  }

  @Override
  public void readListEnd() throws TException {
    concreteProtocol.readListEnd();
  }

  @Override
  public TSet readSetBegin() throws TException {
    return concreteProtocol.readSetBegin();
  }

  @Override
  public void readSetEnd() throws TException {
    concreteProtocol.readSetEnd();
  }

  @Override
  public boolean readBool() throws TException {
    return concreteProtocol.readBool();
  }

  @Override
  public byte readByte() throws TException {
    return concreteProtocol.readByte();
  }

  @Override
  public short readI16() throws TException {
    return concreteProtocol.readI16();
  }

  @Override
  public int readI32() throws TException {
    return concreteProtocol.readI32();
  }

  @Override
  public long readI64() throws TException {
    return concreteProtocol.readI64();
  }

  @Override
  public double readDouble() throws TException {
    return concreteProtocol.readDouble();
  }

  @Override
  public String readString() throws TException {
    return concreteProtocol.readString();
  }

  @Override
  public ByteBuffer readBinary() throws TException {
    return concreteProtocol.readBinary();
  }

  /**
   * @param type Returns the minimum amount of bytes needed to store the smallest possible instance
   *     of TType.
   * @return
   * @throws TException
   */
  @Override
  public int getMinSerializedSize(byte type) throws TException {
    return concreteProtocol.getMinSerializedSize(type);
  }
}
